eox_tenant/signals.py

```diff
@@ -217,3 +231,90 @@ def clear_tenant(sender, request, **kwargs):  # pylint: disable=unused-argument
     Signal: django.core.signals.got_request_exception
     """
     pass
+
+
+def before_task_publish(sender, *args, **kwargs):  # pylint: disable=unused-argument
+    """
+    Dispatched before a celery task is published. Note that this is executed in the process sending the task.
+    See:
+        https://celery.readthedocs.io/en/latest/userguide/signals.html#before-task-publish
+    """
+    headers = kwargs.get('headers')
+    body = kwargs.get('body')
+    try:
+        request_info = body.get('args')[1].get('request_info')
+        headers['eox_tenant_sender'] = request_info.get('host') or None
+    except AttributeError:
+        # This is a async taks generated outside of the scope of a request i.e. CLI task
+        get_host_func = AsyncTaskHandler().get_host_from_task(sender)
+        headers['eox_tenant_sender'] = get_host_func(body)
+
+    headers['X-Forwarded-For'] = 'CRITICAL_PIECE_OF_INFO'
+
+
+def task_prerun(sender, *args, **kwargs):  # pylint: disable=unused-argument
+    """
+    Dispatched before a task is executed.
+    See:
+       https://celery.readthedocs.io/en/latest/userguide/signals.html#task-prerun
+    """
+    headers = sender.request.get('headers', {})
+
+    http_host = headers.get('eox_tenant_sender', None)
+
+    if not http_host:
+        LOG.warning("Could not find the host information for eox_tenant.signals")
+
+    domain = http_host.split(':')[0]
+
+    can_keep = reset_or_keep_settings(domain)
+
+    if can_keep:
+        return
+
+    # Do the update
+    _update_settings(domain)
+
+
+class AsyncTaskHandler(object):
+    """
+    Handler used to get the tenant of an async task outsite of the scope of a request.
+    """
+    TASKS = {
+        "openedx.core.djangoapps.schedules.tasks.ScheduleRecurringNudge": 'get_host_from_siteid'
+    }
+
+    def get_host_from_task(self, sender):
+        """
+        Method used to get the function used to get the tenant host
+        accordingly with type of task.
+        """
+        value = self.TASKS.get(sender, 'default_host')
+        action = getattr(self, value)
+        return action()
+
+    def get_host_from_siteid(self):
+        """
+        Get the tenant domain name using its Site id
+        """
+        def get_host(body):
+            """
+            Get host from siteid.
+            """
+            siteid = body.get('args')[0]
+            site = Site.objects.get(id=siteid)
+            return site.name
+        return get_host
+
+    def default_host(self):
+        """
+        Get default tenant.
+        """
+        def get_host(_body):
+            """
+            Get host using default site id.
+            """
+            siteid = base_settings.SITE_ID
+            site = Site.objects.get(id=siteid)
+            return site.name
+        return get_host
```

Comment:
I believe this is the place where we should be logging the fact that we did not find the tenant and we need to raise the alarms.